psi4/driver/driver_findif.py

```diff
@@ -68,7 +68,9 @@ def _displace_cart(mol, geom, salc_list, i_m, step_size):
     # This for loop and tuple unpacking is why the function can handle
     # an arbitrary number of SALCs.
     for salc_index, disp_steps in i_m:
-        for component in salc_list[salc_index]:
+        # * Python error if iterate through `salc_list`
+        for i in range(len(salc_list[salc_index])):
```

Comment:
I still don't get why this is a problem in Jupyter but not from (the same) python shell.

Reply:
I'm with Peter on this one. Did I just make a mistake when I did the PyBinding of `CdSalcList.__iter__`?

Reply:
I'm getting it now command-line, too, and I hadn't seen it that way before. I suspect something awry btwn python version, pb11, and maybe the binding, too. Possibly it could be fixed by through pybinding, but I don't see an obvious suggestion.

Reply:
In the xtensor branch I've implemented the quadruplet as follows:
```c++
cls.def("__getitem__", py::overload_cast<size_t>(&Class::operator[]), "Return block at given irrep", "h"_a,
            py::is_operator(), py::return_value_policy::reference_internal);
cls.def("__setitem__",
            [](Class& obj, size_t h, const xt::pytensor<T, Rank>& block) {
                if (h >= obj.nirrep()) throw py::index_error();
                obj[h] = block;
            },
            "h"_a, "block"_a, "Set block at given irrep", py::is_operator());
cls.def("__len__", &Class::nirrep);
cls.def("__iter__", [](const Class& obj) { return py::make_iterator(obj.begin(), obj.end()); },
            py::keep_alive<0, 1>() /* Essential: keep object alive while iterator exists */);
```
Don't know if it could be related or not, but I pass the beginning and ending of the collection to `py::make_iterator` in the implementation of `__iter__`

Reply:
fwiw, I tried the begin/end below. It compiles but the `__next__ of PyCapsule` error resumes.
```
         .def(
-            "__iter__", [](const CdSalc& salc) { return py::make_iterator(salc.get_components()); },
+            "__iter__", [](const CdSalc& salc) { return py::make_iterator(salc.get_components().begin(), salc.get_components().end()); },
             py::keep_alive<0, 1>());
 
     py::class_<CdSalcList, std::shared_ptr<CdSalcList>>(
@@ -706,13 +706,25 @@ void export_mints(py::module& m) {
         .def("__getitem__", [](const CdSalcList& salclist, size_t i) { return salclist[i]; })
         .def("__len__", [](const CdSalcList& salclist) { return salclist.ncd(); })
         .def(
-            "__iter__", [](const CdSalcList& salclist) { return py::make_iterator(salclist.get_salcs()); },
+            "__iter__", [](const CdSalcList& salclist) { return py::make_iterator(salclist.get_salcs().begin(), salclist.get_salcs().end()); },
             py::keep_alive<0, 1>())
```

Reply:
Problem showed up again for `core.set_local_option`. I've found that the appearance of the `PyCapsule ... returned NULL` is dependent on what other packages are installed. In particular, something in the QCFractal stack triggers it. The transaction below removed the problem. Still needs isolating.

```
The following packages will be REMOVED:

  alembic-1.4.0-py_0
  bcrypt-3.1.7-py38h516909a_0
  libpq-11.2-h20c2e04_0
  mako-1.1.0-py_0
  markupsafe-1.1.1-py38h516909a_0
  postgresql-11.2-h20c2e04_0
  psycopg2-2.8.4-py38h1ba5d50_0
  python-editor-1.0.4-py_0
  sqlalchemy-1.3.13-py38h516909a_0
  tornado-6.0.3-py38h516909a_2
  tqdm-4.42.1-py_0

The following packages will be UPDATED:

  boost-cpp          conda-forge::boost-cpp-1.69.0-h11c811~ --> pkgs/main::boost-cpp-1.71.0-h7b6447c_0
  ca-certificates    conda-forge::ca-certificates-2019.11.~ --> pkgs/main::ca-certificates-2020.1.1-0
  cffi                                1.13.2-py38h2e261b9_0 --> 1.14.0-py38h2e261b9_0
  ipython                             7.11.1-py38h39e3cac_0 --> 7.12.0-py38h5ca1d4c_0
  jedi                                        0.15.2-py38_0 --> 0.16.0-py38_0
  krb5                                    1.16.4-h173b8e3_0 --> 1.17.1-h173b8e3_0
  libcurl                                 7.67.0-h20c2e04_0 --> 7.68.0-h20c2e04_0
  more-itertools                                 8.0.2-py_0 --> 8.2.0-py_0
  ncurses                                    6.1-he6710b0_1 --> 6.2-he6710b0_0
  numpy-base                          1.18.1-py38hde5b4d6_0 --> 1.18.1-py38hde5b4d6_1
  openssl            conda-forge::openssl-1.1.1d-h516909a_0 --> pkgs/main::openssl-1.1.1d-h7b6447c_4
  packaging                                       20.0-py_0 --> 20.1-py_0
  parso                                          0.5.2-py_0 --> 0.6.1-py_0
  pexpect                                      4.7.0-py38_0 --> 4.8.0-py38_0
  pip                                         19.3.1-py38_0 --> 20.0.2-py38_1
  prompt_toolkit                                 3.0.2-py_0 --> 3.0.3-py_0
  regex                            2019.12.9-py38h7b6447c_0 --> 2020.2.20-py38h7b6447c_0
  setuptools                                  44.0.0-py38_0 --> 45.2.0-py38_0
  six                                         1.13.0-py38_0 --> 1.14.0-py38_0
  sqlite                                  3.30.1-h7b6447c_0 --> 3.31.1-h7b6447c_0
  typed-ast                            1.4.0-py38h7b6447c_0 --> 1.4.1-py38h7b6447c_0
  urllib3                                     1.25.7-py38_0 --> 1.25.8-py38_0
  wcwidth            pkgs/main/linux-64::wcwidth-0.1.7-py3~ --> pkgs/main/noarch::wcwidth-0.1.8-py_0
  wheel                                       0.33.6-py38_0 --> 0.34.2-py38_0

The following packages will be SUPERSEDED by a higher-priority channel:

  brotli             conda-forge::brotli-1.0.7-he1b5a44_10~ --> pkgs/main::brotli-1.0.7-he6710b0_0
  c-ares             conda-forge::c-ares-1.15.0-h516909a_1~ --> pkgs/main::c-ares-1.15.0-h7b6447c_1001
  certifi                                       conda-forge --> pkgs/main
  double-conversion  conda-forge::double-conversion-3.1.5-~ --> pkgs/main::double-conversion-3.1.5-he6710b0_1
  gflags             conda-forge::gflags-2.2.2-he1b5a44_10~ --> pkgs/main::gflags-2.2.2-he6710b0_0
  glog                   conda-forge::glog-0.4.0-he1b5a44_1 --> pkgs/main::glog-0.4.0-he6710b0_0
  libevent           conda-forge::libevent-2.1.10-h72c5cf5~ --> pkgs/main::libevent-2.1.8-h1ba5d50_0
  lz4-c              conda-forge::lz4-c-1.8.3-he1b5a44_1001 --> pkgs/main::lz4-c-1.8.1.2-h14c3975_0
  mypy_extensions                               conda-forge --> pkgs/main
  pathspec                                      conda-forge --> pkgs/main
  re2                conda-forge::re2-2020.01.01-he1b5a44_0 --> pkgs/main::re2-2019.08.01-he6710b0_0
  readline           conda-forge::readline-7.0-hf8c457e_10~ --> pkgs/main::readline-7.0-h7b6447c_5
  snappy               conda-forge::snappy-1.1.8-he1b5a44_1 --> pkgs/main::snappy-1.1.7-hbae5bb6_3
  tk                      conda-forge::tk-8.6.10-hed695b0_0 --> pkgs/main::tk-8.6.8-hbc83047_0
  uriparser          conda-forge::uriparser-0.9.3-he1b5a44~ --> pkgs/main::uriparser-0.9.3-he6710b0_1

The following packages will be DOWNGRADED:

  arrow-cpp                           0.15.1-py38h248a92f_5 --> 0.15.1-py38h7cd5009_5
```